Abstract predicates

vl0w opened the pull request:
This PR introduces *abstract predicates*, which are predicates which are declared on a trait:
```rust
trait SomeTrait {
    // Define an abstract predicate
    predicate! {
        fn pred(&self) -> bool;
    }
}
struct SomeStruct { ... }
#[refine_trait_spec]
impl SomeTrait for SomeStruct {
    // Implement the abstract predicate
    predicate! {
        fn pred(&self) -> bool {
            ...
        }
    }
}
```

These abstract predicates can then be used in spec code:
```rust
#[requires(s.pred())]
#[ensures(s.pred())]
fn foo(s: SomeStruct) { ... }
```
And even in a generic setting:
```rust
#[requires(s.pred())]
#[ensures(s.pred())]
fn bar<T: SomeTrait>(s: T) { ... }
```

An abstract predicate (like in `bar` above) is always encoded as a pure bodyless Viper function. Thus, abstract predicates can never have a body (i.e. a default implementation on the trait method). This restriction also allows us to not need to worry about refinement.

## Other notable changes
- `#[refine_trait_spec]` now supports implementing a local trait on a non-local type. This allows for implementing abstract predicates on non-local types.
    Implementation detail: Spec functions in `#[refine_trait_spec] impl A for B` are not anymore generated in an `impl B` block, but are free-standing functions now.

# Work to do
- [x] Parsing
- [x] Ensure that abstract predicates are not used in non-spec code
- [x] Ensure abstract predicates don't have bodies
- [x] Ensure that we don't declare predicates in extern specs
- [x] Support extern types in `#[refine_trait_spec]`
- [x] Resolve all `TODO hansenj` and `TODO abstract-predicates`
- [x] Adapt to new `ProcedureSpecificationKind` and handle abstract predicates
- [x] Testing (Parsing, Validation, Usage)

vl0w committed:
Parsing of abstract predicates

vl0w committed:
Error handling for abstract predicates

- Must not have bodies
- May not appear in non-spec code

vl0w committed:
Raise error when declaring predicates in external specs

vl0w committed:
Support trait spec refinements for non-local types

Spec functions in `#[refine_trait_spec]` will longer be attached to an impl block, they are free functions now.

vl0w committed:
Parse abstract predicates in SpecCollector

Additionally, if there is no specification for a procedure we default the ProcedureSpecificationKind to impure (never create an "empty" kind)

vl0w committed:
Remove test

vl0w committed:
Adapt abstract predicate validation to new spec checker structure

vl0w committed:
Start encoding abstract predicates as bodyless functions

vl0w committed:
Merge branch 'master' into abstract-predicates

# Conflicts:
#	prusti-interface/src/specs/typed.rs
#	prusti-interface/src/utils.rs
#	prusti-specs/src/extern_spec_rewriter/common.rs
#	prusti-specs/src/extern_spec_rewriter/traits.rs
#	prusti-specs/src/lib.rs
#	prusti-tests/tests/parse/ui/predicates-visibility.stdout
#	prusti-tests/tests/parse/ui/predicates.stdout
#	prusti-tests/tests/verify/ui/predicate.stdout
#	prusti-viper/src/encoder/mir/pure/pure_functions/interface.rs
#	prusti-viper/src/encoder/mir/specifications/interface.rs

vl0w committed:
Merge fixes

vl0w committed:
Resolve ghost constraints even in a generic setting

i.e. when encoding a function which is generic and we resolve the ghost constraint over some of this generic's function

vl0w committed:
Small refactoring

vl0w committed:
Parse extern spec spec functions eagerly + refactorings

vl0w committed:
Add tests

vl0w committed:
Copy generics when refining a trait spec

Aurel300 commented on prusti-interface/src/specs/checker/predicate_checks.rs:
```diff
@@ -114,16 +144,33 @@ impl<'v, 'tcx: 'v> NonSpecExprVisitor<'tcx> for CheckPredicatesVisitor<'tcx> {
-            let def_id = ex.hir_id.owner;
-            if self.tcx.is_mir_available(def_id) && !self.tcx.is_constructor(def_id.to_def_id()) {
-                let res = self.tcx.typeck(def_id).qpath_res(path, ex.hir_id);
+            if self.tcx.is_mir_available(owner_def_id) && !self.tcx.is_constructor(owner_def_id.to_def_id()) {
```
Why are we checking for `is_constructor` (and `is_mir_available`) here and in line 165?

vl0w replied:
Idk to be honest. These checks were already there and I did not question them. Let's see if the build fails.

Aurel300 commented on prusti-specs/src/common.rs:
````diff
@@ -1,29 +1,338 @@
+    /// }
+    /// ```
+    pub(crate) trait RewritableReceiver {
+        fn rewrite_receiver<T: ToTokens>(&mut self, new_ty: &T);
````
Can the `T` here be something specific, e.g. `TypePath`? Is there a use for providing different tokenisable types?

vl0w replied:
I think the original idea of this was because the implementation is purely token based, i.e. `TypePath` would be tokenized anyway. But I think this is a good idea, the interface should be designed according to client's needs, not the implementation.

vl0w replied:
And actually, I think this was not considered during my refactoring (of the refactoring). `T` makes no sense anymore.

Aurel300 commented on prusti-specs/src/common.rs:
```diff
@@ -1,29 +1,338 @@
+            }));
+            parse_quote_spanned! {tokens_span=>
+            #rewritten
+        }
```
Indent mishap.

Aurel300 commented on prusti-specs/src/common.rs:
```diff
@@ -1,29 +1,338 @@
+    let generics_source = source.generics();
+
+    generics_target.params.extend(generics_source.params.clone());
+    if let (Some(target_where), Some(source_where)) =
```
What if `generics_target.where_clause` is `None` but `generics_source.where_clause` is `Some(_)`? Then `generics_target` should become `Some(_)`, right?

vl0w replied:
Indeed! Also, type params could be duplicated. I fixed this and added some tests for it.

Aurel300 commented on prusti-specs/src/extern_spec_rewriter/common.rs:
```diff
@@ -1,27 +1,138 @@
+    stub_method.rewrite_receiver(&self_type_path);
+
+    // Set span of generated method to externally specified method for better error reporting
+    syn::visit_mut::visit_impl_item_method_mut(&mut SpanOverrider::new(method_sig_span), &mut stub_method);
```
Which spans will this override at this point? Is it still necessary?

vl0w replied:
I re-added this line because some tests were failing.

Aurel300 commented on prusti-specs/src/extern_spec_rewriter/common.rs:
```diff
@@ -1,27 +1,138 @@
-            _ => token.into_token_stream(),
-        }))
+        }
+        args
```
`Punctuated` has [`FromIterator`](https://docs.rs/syn/latest/syn/punctuated/struct.Punctuated.html#impl-FromIterator%3CT%3E), so all this should be expressible with an iterator chain + `map` + `collect`. The `push_punct` in particular seems unneccessary if it is not correctly spanned anyway.

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -394,43 +409,52 @@ pub fn refine_trait_spec(_attr: TokenStream, tokens: TokenStream) -> TokenStream
+                let predicate = match parsed_predicate {
+                    ParsedPredicate::Impl(predicate) => predicate,
+                    _ => unreachable!(),
+                };
```
Can we use our `force_match!` macro for these?

Aurel300 commented on prusti-tests/tests/typecheck/fail/abstract-predicates/invalid-sig-1.rs:
```diff
@@ -0,0 +1,22 @@
+impl SomeTrait for SomeStruct {
+    predicate! {
+        fn foo(&self) -> i32 {
+            42 //~ ERROR: mismatched types
```
Why this error?

vl0w replied:
I created a UI test, the error message is better there.

Aurel300 commented on prusti-tests/tests/verify_overflow/pass/trait-contracts-refinement/refine-non-local-type-with-generics.rs:
```diff
@@ -0,0 +1,35 @@
+
+    #[pure]
+    #[ensures(self.is_some() == ! result)]
+    #[ensures(matches ! (* self, None) == result)]
```
Is the funky spacing here intentional?

vl0w replied:
No :)

Aurel300 commented on prusti-viper/src/encoder/mir/pure/pure_functions/interface.rs:
```diff
@@ -275,15 +276,19 @@ impl<'v, 'tcx: 'v> PureFunctionEncoderInterface<'v, 'tcx>
+                    self.is_trusted(proc_def_id, Some(substs)),
+                    self.get_proc_kind(proc_def_id, Some(substs)),
+                ) {
+                    // Predicate bodies are encoded even though the function is trusted
```
This seems strange: do we need to support this? Can we not error out in parsing and/or not mark a predicate as trusted?

vl0w replied:
My change there resulted from existing code, so I did not try to change existing behavior.
I double-checked: We can not mark a predicate as trusted.

vl0w replied:
My question then is how should the encode work if we had trusted predicates? Should they be encoded as bodyless functions (like trusted pure functions) or should the body still be encoded? Before my change, a predicate would not have been encoded by a bodyless function.

Aurel300 replied:
I guess they would be abstract/bodyless, yes… The motivation for predicates was to support Prusti syntax inside methods, but with a trusted method the body should be ignored anyway. So in the end, the only thing a trusted predicate does vs a trusted pure method is that it cannot be used in non-spec code. I guess that would be the case if you could declare these in an extern spec but maybe that's a stretch.

So, if we cannot mark a predicate as trusted anyway (not even via extern specs), can this code be simplified any (or at least the comment changed)?

vl0w replied:
Yes, we also cannot add a `trusted` flag via extern spec, even though the error is not really helpful ("cannot use predicate in non-spec code" because of the expansion of the extern spec).

If the general behavior is to always encode a trusted function as bodyless I will adjust it.

vl0w replied:
My naive understanding and refactoring caused `ui/predicate.rs` to fail.

I (maybe) understand this better now: Predicate functions *are* trusted. In contrast to pure functions, we cannot add a functional spec to the predicate, the predicate *is* the spec. Therefore, in the pure function encoder, we have to check first whether a function is a predicate to encode it, because they will otw. not do anything at all.

Does my explanation make sense? I think I can adjust the comment in the code a bit to clarify.

Aurel300 replied:
This indirection of a spec item in a separate method is an implementation detail. I don't want to conflate this with `trusted` methods in the code or comments too much.

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/interface.rs:
```diff
@@ -63,26 +65,29 @@ impl<'tcx> SpecQuery<'tcx> {
 }
 
 pub(crate) trait SpecificationsInterface<'tcx> {
+    // TODO abstract-predicates: Maybe this should be deleted (and ProcedureSpecificationKind::is_pure)
```
I agree, it might be confusing that predicates also return `true` here. How many usages are there?

vl0w replied:
I found 8 in total.
We maybe could just rename the method to `must_be_encoded_as_pure` or something similar, afaik this is how we interact with this method.

vl0w committed:
Remove generic from RewritableReceiver

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Refactor impl of MethodParamsAsCallArguments

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Fix bugs in merge_generics

vl0w committed:
Remove funky spacings

vl0w committed:
Better check for error in UI test

vl0w committed:
Refactoring: Use force_matches!

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Remove (unnecessary?) check for mir availability in predicate_checks.rs

Aurel300 commented on prusti-specs/src/common.rs:
```diff
@@ -1,29 +1,358 @@
+                    GenericParam::Type(tp) => tp.ident.eq(&type_param_source.ident),
+                    _ => false
+                }
+                ).map(|gp| force_matches!(gp, GenericParam::Type(tp) => tp));
```
This sounds like you want a `HashSet` of the identifiers already inserted...

vl0w replied:
It screams for `HashABC`s ^^
I refactored it again. Maybe it could be done more elegantly?

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -394,43 +409,49 @@ pub fn refine_trait_spec(_attr: TokenStream, tokens: TokenStream) -> TokenStream
+                let mut spec_function = match predicate.spec_function {
+                    syn::Item::Fn(item_fn) => item_fn,
+                    _ => unreachable!(),
+                };
```
Is this also not a `force_matches!`?

Aurel300 commented on prusti-tests/tests/typecheck/ui/abstract-predicates/invalid-sig-1.stderr:
```diff
@@ -0,0 +1,9 @@
+error[E0308]: mismatched types
+  --> $DIR/invalid-sig-1.rs:15:13
+   |
+15 |             42 //~ ERROR: mismatched types
```
This still doesn't explain what this error is. Are predicates supposed to only return `bool`? I don't think this should be the case. (If it were, the error here should highlight the return type, not the expression.) I'm guessing this error is raised because of a `(... : bool)` type check expression, which should not exist for predicate spec items (but should for pre/postconditions).

vl0w replied:
I believed that predicates are meant to return `bool`. However, the user docs don't mention that and I think my assumption was wrong after all.
However, this PR doesn't change the encoding/behavior of predicates at all.
I think if we want to fix this we should do it in a separate PR then and I will delete this test.

vl0w replied:
I deleted the test

vl0w committed:
Undo: Remove (unnecessary?) check for mir availability in predicate_checks.rs

vl0w committed:
Always encode trusted functions as bodyless functions in pure function encoder

vl0w committed:
Delete wrong test

vl0w committed:
Refactoring of merge_generics

vl0w committed:
Encode predicate with bodies, even though they are trusted

vl0w committed:
Handle generics for predicate macro in #[refine_trait_spec] correctly

vl0w committed:
Use Vec instead of HashSet in merge_generics to have a deterministic order

vl0w committed:
Make predicates non-trusted

The encoder will still not process predicates in its queue, just handle them separately

vl0w committed:
Add new parameters in merge_generics as a prefix
